tests/lib/Controller/Oauth2ControllerTest.php

```diff
@@ -0,0 +1,300 @@
+        $this->requestMock->query = $this->createQueryMock([]);
+
+        /** @psalm-suppress UndefinedMethod,MixedMethodCall */
+        $this->controller
+            ->expects($this->once())
+            ->method('handleError');
+
+        $this->controller->linkback($this->requestMock);
+    }
+
+    public function testLinkbackWithIdentityProviderException(): void
+    {
+        $this->createControllerMock(['parseRequest', 'getSourceService', 'handleError', 'getHttp', 'parseError']);
+        $this->requestMock->query = $this->createQueryMock(['code' => 'validCode']);
+
+        $this->oauth2Mock->expects($this->once())
+            ->method('finalStep')
+            ->willThrowException(new IdentityProviderException('Error Message', 0, ['body' => 'error body']));
+
+        $this->expectException(AuthSource::class);
+
+        $this->controller->linkback($this->requestMock);
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function testHandleErrorWithConsentedError(): void
+    {
+        $this->createControllerMock(['parseRequest', 'getSourceService', 'getHttp', 'parseError']);
+        $this->controller->getSource()->method('getAuthId')->willReturn('authId');
+        $this->controller
+            ->getSource()
+            ->method('getConfig')
+            ->willReturn(new class (['useConsentErrorPage' => true], '') extends Configuration {
+                public function getOptionalBoolean($name, $default): bool
+                {
+                    if (!$this->hasValue($name) && isset($default)) {
+                        return filter_var($default, FILTER_VALIDATE_BOOLEAN);
+                    }
+                    return filter_var($this->getValue($name), FILTER_VALIDATE_BOOLEAN);
+                }
+            });
+
+        $this->requestMock->query = $this->createQueryMock(
+            ['error' => 'invalid_scope', 'error_description' => 'Invalid scope']
+        );
+
+        $this->controller->method('parseError')
+            ->with($this->requestMock)
+            ->willReturn(['invalid_scope', 'Invalid scope']);
+
+        $this->controller->method('getHttp')->willReturn($this->httpMock);
+
+        $this->httpMock->expects($this->once())
+            ->method('redirectTrustedURL')
+            ->with('http://localhost/module.php/authoauth2/errors/consent');
+
+        $this->controller->handleError($this->controller->getSource(), $this->stateMock, $this->requestMock);
+    }
+
+    public function testHandleErrorWithConsentErrorAtLegacyRoute(): void
+    {
+        $this->createControllerMock(['parseRequest', 'getSourceService', 'getHttp', 'parseError']);
+        $this->controller->getSource()->method('getAuthId')->willReturn('authId');
+        $this->controller
+            ->getSource()
+            ->method('getConfig')
+            ->willReturn(new class ([
+                'useConsentErrorPage' => true,
+                'useLegacyRoutes' => true
+            ], '') extends Configuration {
```

Comment:
what was the need for subclassing a configuration object?

Reply:
When i first wrote the test i wanted to override the `public function getOptionalBoolean($name, $default)` function and always `return true`. After the latest changes the test became more complex and i had to add more logic. With its current form overriding the function is no longer needed. Thank you for suggesting simplifying this one.

Reply:
Done